packages/node-ws/src/index.ts

```diff
@@ -0,0 +1,115 @@
+import { Buffer } from 'buffer'
+import type { Server } from 'node:http'
+import type { Http2SecureServer, Http2Server } from 'node:http2'
+import type { Hono } from 'hono'
+import { createMiddleware } from 'hono/factory'
+import type { UpgradeWebSocket, WSContext } from 'hono/ws'
+import { WebSocketServer } from 'ws'
+
+export interface NodeWebSocket {
+  upgradeWebSocket: UpgradeWebSocket
+  injectWebSocket(server: Server | Http2Server | Http2SecureServer): void
+}
+export interface NodeWebSocketInit {
+  app: Hono
+  baseUrl?: string | URL
+}
+
+/**
+ * Extended for telling WebSocket
+ * @internal
+ */
+class WSResponse extends Response {
+  readonly wss: WebSocketServer
+  constructor(wss: WebSocketServer) {
+    super()
+    this.wss = wss
+  }
+}
+
+/**
+ * Create WebSockets for Node.js
+ * @param init Options
+ * @returns NodeWebSocket
+ */
+export const createNodeWebSocket = (init: NodeWebSocketInit): NodeWebSocket => {
+  return {
+    injectWebSocket(server) {
+      ;(server as Server).on('upgrade', async (request, socket, head) => {
+        const url = new URL(request.url ?? '/', init.baseUrl ?? 'http://localhost')
+        const headers = new Headers()
+        for (const key in request.headers) {
+          const value = request.headers[key]
+          if (!value) {
+            continue
+          }
+          headers.append(key, Array.isArray(value) ? value[0] : value)
+        }
+        const res = (await init.app.request(url, {
+          headers: headers,
+        })) as Response | WSResponse
+        if (!(res instanceof WSResponse)) {
+          socket.destroy()
+          return
+        }
+        res.wss.handleUpgrade(request, socket, head, (ws) => {
+          res.wss.emit('connection', ws, request)
+        })
+      })
+    },
+    upgradeWebSocket: (createEvents) =>
+      createMiddleware(async (c, next) => {
```

Comment:
It's better to make it a named function, see:

https://github.com/honojs/hono/pull/2615#discussion_r1591717786

Reply:
I fixed at https://github.com/honojs/middleware/pull/503/commits/36483096100b797ecc25d501cd77f633d702fe32.